2011-08-22  Alan Modra  <amodra@gmail.com>

	* ld-powerpc/tocopt4a.s, * ld-powerpc/tocopt4b.s,
	* ld-powerpc/tocopt4.d: New test.
	* ld-powerpc/powerpc.exp: Run it.

2011-08-19  Alan Modra  <amodra@gmail.com>

	* elf64-ppc.c (ppc64_elf_edit_toc): Ignore can_optimize bit if
	we haven't seen expected -mcmodel=medium/large code relocs.

Index: binutils-2.20.51.0.2/bfd/elf64-ppc.c
===================================================================
--- binutils-2.20.51.0.2.orig/bfd/elf64-ppc.c
+++ binutils-2.20.51.0.2/bfd/elf64-ppc.c
@@ -8196,7 +8196,7 @@ ppc64_elf_edit_toc (struct bfd_link_info
 		some_unused = 1;
 	      last = 0;
 	    }
-	  else if (*drop)
+	  else if (*drop & ref_from_discarded)
 	    {
 	      some_unused = 1;
 	      last = ref_from_discarded;
Index: binutils-2.20.51.0.2/ld/testsuite/ld-powerpc/powerpc.exp
===================================================================
--- binutils-2.20.51.0.2.orig/ld/testsuite/ld-powerpc/powerpc.exp
+++ binutils-2.20.51.0.2/ld/testsuite/ld-powerpc/powerpc.exp
@@ -179,6 +179,8 @@ set ppc64elftests {
 	{{ld tocopt2.out} {objdump -s tocopt2.d}} "tocopt2"}
     {"TOC opt3" "-melf64ppc -no-keep-memory --defsym x=2" "-a64"  {tocopt3.s}
 	{{objdump -s tocopt3.d}} "tocopt3"}
+    {"TOC opt4" "-melf64ppc -no-keep-memory --defsym x=2" "-a64"
+        {tocopt4a.s tocopt4b.s} {{objdump -s tocopt4.d}} "tocopt4"}
 }
 
 
Index: binutils-2.20.51.0.2/ld/testsuite/ld-powerpc/tocopt4.d
===================================================================
--- /dev/null
+++ binutils-2.20.51.0.2/ld/testsuite/ld-powerpc/tocopt4.d
@@ -0,0 +1,7 @@
+
+.*:     file format .*
+
+Contents of section \.text:
+ 100000b0 e9298000 .*
+Contents of section \.got:
+ 100100b8 00000000 00000002 .*
Index: binutils-2.20.51.0.2/ld/testsuite/ld-powerpc/tocopt4a.s
===================================================================
--- /dev/null
+++ binutils-2.20.51.0.2/ld/testsuite/ld-powerpc/tocopt4a.s
@@ -0,0 +1,4 @@
+ .globl _start
+ .text
+_start:
+ ld 9,xt@toc(9)
Index: binutils-2.20.51.0.2/ld/testsuite/ld-powerpc/tocopt4b.s
===================================================================
--- /dev/null
+++ binutils-2.20.51.0.2/ld/testsuite/ld-powerpc/tocopt4b.s
@@ -0,0 +1,17 @@
+ .section .toc,"aw"
+ .globl xt
+xt:
+ .quad x
+#
+# This testcase will fail with a warning "xt defined on removed toc entry"
+# if a large-toc access like the following is added to this file, because
+# toc analysis only considers toc accesses from the current object file.
+# The small-toc access from tocopt4a.s doesn't cause xt entry to be marked
+# !can_optimize.  The testcase only passes because toc analysis considers
+# *no* access from the current file as being sufficiently unusual to
+# warrant keeping the toc entry.  So, if you use global symbols on toc
+# entries, don't mix code models.
+#
+# .text
+# addis 9,2,xt@toc@ha
+# ld 9,xt@toc@l(9)
